src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,255 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+/**
+ * Base class for flow scanners, which offers basic methods and stubs for algorithms
+ * Scanners store state internally, and are not thread-safe but are reusable
+ * Scans/analysis of graphs is implemented via internal iteration to allow reusing algorithm bodies
+ * However internal iteration has access to additional information
+ *
+ * Provides 4 sets of common APIs to use, in decreasing expressiveness and increasing genericness:
+ *   - findFirst - find first node match a predicate
+ *   - filteredNodes - return a collection of nodes filtered by a predicate, between heads(inclusive) and blackList (exclusive)
+ *   - visitor - call a visitor on each node we encounter
+ *   - Iterator/filterator based: FlowNode-by-FlowNode walking, after setup with head/blacklist
+ *
+ *  All operations can start with one or more "head" FlowNodes, and walk back from there, stopping when we hit blackList nodes
+ *
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public abstract class AbstractFlowScanner implements Iterable <FlowNode>, Filterator<FlowNode> {
+
+    protected FlowNode _current;
+
+    protected FlowNode _next;
+
+    protected Collection<FlowNode> _blackList = Collections.EMPTY_SET;
+
+    /** Convert stop nodes to a collection that can efficiently be checked for membership, handling null if needed */
+    @Nonnull
+    protected Collection<FlowNode> convertToFastCheckable(@CheckForNull Collection<FlowNode> nodeCollection) {
+        if (nodeCollection == null || nodeCollection.size()==0) {
+            return Collections.EMPTY_SET;
+        } else  if (nodeCollection.size() == 1) {
+            return Collections.singleton(nodeCollection.iterator().next());
+        } else if (nodeCollection instanceof Set) {
+            return nodeCollection;
+        }
+        return nodeCollection.size() > 5 ? new HashSet<FlowNode>(nodeCollection) : nodeCollection;
+    }
+
+    /**
+     * Set up for iteration/analysis on a graph of nodes, initializing the internal state
+     * @param heads The head nodes we start walking from (the most recently executed nodes,
+     *              i.e. FlowExecution.getCurrentHeads()
+     * @param blackList Nodes that we cannot visit or walk past (useful to limit scanning to only nodes after a specific point)
+     * @return True if we can have nodes to work with, otherwise false
+     */
+    public boolean setup(@CheckForNull Collection<FlowNode> heads, @CheckForNull Collection<FlowNode> blackList) {
+        if (heads == null || heads.size() == 0) {
+            return false;
+        }
+        Collection<FlowNode> fastEndNodes = convertToFastCheckable(blackList);
+        HashSet<FlowNode> filteredHeads = new HashSet<FlowNode>(heads);
```

Comment:
Perhaps you want to use `LinkedHashSet`? I presume there are some cases where the order in which heads were specified could affect the result, right?